server/src/grpc/remote_engine_service/mod.rs

```diff
@@ -16,37 +16,95 @@ use ceresdbproto::{
     storage::{arrow_payload, ArrowPayload},
 };
 use common_types::record_batch::RecordBatch;
-use futures::stream::{self, BoxStream, StreamExt};
+use futures::stream::{self, BoxStream, FuturesUnordered, StreamExt};
 use generic_error::BoxError;
 use log::{error, info};
 use proxy::instance::InstanceRef;
 use query_engine::executor::Executor as QueryExecutor;
 use snafu::{OptionExt, ResultExt};
 use table_engine::{
-    engine::EngineRuntimes, remote::model::TableIdentifier, stream::PartitionedStreams,
-    table::TableRef,
+    engine::EngineRuntimes,
+    predicate::PredicateRef,
+    remote::model::TableIdentifier,
+    stream::PartitionedStreams,
+    table::{ReadOrder, TableRef},
 };
 use time_ext::InstantExt;
 use tokio::sync::mpsc;
 use tokio_stream::wrappers::ReceiverStream;
 use tonic::{Request, Response, Status};
 
-use crate::grpc::{
-    metrics::{
-        REMOTE_ENGINE_GRPC_HANDLER_COUNTER_VEC, REMOTE_ENGINE_GRPC_HANDLER_DURATION_HISTOGRAM_VEC,
+use crate::{
+    dedup_requests::{RequestNotifiers, RequestResult},
+    grpc::{
+        metrics::{
+            REMOTE_ENGINE_GRPC_HANDLER_COUNTER_VEC,
+            REMOTE_ENGINE_GRPC_HANDLER_DURATION_HISTOGRAM_VEC,
+        },
+        remote_engine_service::error::{ErrNoCause, ErrWithCause, Result, StatusCode},
     },
-    remote_engine_service::error::{ErrNoCause, ErrWithCause, Result, StatusCode},
 };
 
-mod error;
+pub mod error;
 
 const STREAM_QUERY_CHANNEL_LEN: usize = 20;
 const DEFAULT_COMPRESS_MIN_LENGTH: usize = 80 * 1024;
 
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct RequestKey {
+    table: String,
+    predicate: PredicateRef,
+    projection: Option<Vec<usize>>,
+    order: ReadOrder,
+}
+
+impl RequestKey {
+    pub fn new(
+        table: String,
+        predicate: PredicateRef,
+        projection: Option<Vec<usize>>,
+        order: ReadOrder,
+    ) -> Self {
+        Self {
+            table,
+            predicate,
+            projection,
+            order,
+        }
+    }
+}
+
+struct RequestNotifierGuard<F: FnOnce()>(Option<F>);
```

Comment:
Maybe it'll be better if we put it like:
```suggestion
struct ExecutionGuard<F> {
   do: F,
   cancelled: bool,
}

impl<F> for ExecutionGuard<F> {
   fn cancel(&mut self) {
       self.cancelled = true;
   }
}

impl<F: FnOnce()> Drop for ExecutionGuard<F> {...}
```